doc/PluginSystem.md

```diff
@@ -0,0 +1,152 @@
+
+```javascript
+'use strict';
+
+class HelloWorld {
+  constructor() {
+    this.commands = {
+      deploy: {
+        lifeCycleEvents: [
+          'resources',
+          'functions'
+        ]
+      },
+    };
+  }
+}
+```
+
+We automatically put the name of the command in front of lifecycle events when they are used for hooks.
+So in a hook the following syntax needs to be used.
+
+* ***CommandName:LifecycleEventName***
+
+Which would be ***deploy:resources*** in a hook definition (which we will show in more detail below).
+
+In addition to the lifecycle events defined here we will create 2 additional events for each:
+* ***before:CommandName:LifecycleEventName***
+* ***after:CommandName:LifecycleEventName***
+
+Following the above example we’ll have these lifecycle events:
+
+* ***before:deploy:resources***
+* ***deploy:resources***
+* ***after:deploy:resources***
+* ***before:deploy:functions***
+* ***deploy:functions***
+* ***after:deploy:functions***
+
+These names will be used as hooks to include plugin logic. This allows to set up lifecycle events with generic names, but still make sure they are only executed for specific commands.
+
+You can also nest commands, e.g. if you want to provide a command ```serverless deploy single``` you can simply nest commands as defined in the following example. Those nested commands have their own lifecycle events and do not inherit them from their parents.
+
+```javascript
+constructor() {
+  this.commands = {
+  deploy: {
+    lifeCycleEvents: [
+      'resources',
+      'functions'
+    ],
+    commands: {
+      single: {
+        lifeCycleEvents: [
+          'resources',
+          'functions'
+        ],
+      },
+    },
+  },
+};
+```
+
+## Hook
+Hooks allow to connect specific lifecycle events to functions in a Plugin. In the constructor of your class you define a *hooks* variable that the Plugin System will use once a specific command is running. Any hook can bind to any lifecycle event from any command, not just from commands that the same plugin provides.
+
+This allows to extend any command with additional functionality.
+
+```javascript
+'use strict';
+
+class Deploy {
+  constructor() {
+    this.commands = {
+      deploy: {
+        lifeCycleEvents: [
+          'resources',
+          'functions'
+        ]
+      },
+    };
+
+    this.hooks = {
+      'before:deploy:resources': this.beforeDeployResources,
+      'deploy:resources': this.deployResources,
+      'after:deploy:functions': this.afterDeployFunctions
+    };
+  }
+
+  beforeDeployResources() {
+    console.log('Before Deploy Resources');
```

Comment:
Can the hook implementations decide by themselves if they act synchronously or if they return a promise in case they have to call asynchronous functionality?